### ISSUE TITLE
[reflections] modular programming

### ISSUE THREAD

**Issue by atesin** [nesbox/TIC-80#2668]
Hello, I hope this message were understood, I know it is a bit long so for my convenience I wrote it in Spanish (I am Chilean) and translated mostly with Google Translate, I hope @nesbox and @oblerion read it since it has to do with TIC-80 and code development

Let's be honest, I love TIC-80, it looks like the Atari I had when I was a kid, and although it is used by many people with low-level programming knowledge (assembler, memory maps, etc.), because of its simplicity it has also been chosen by more beginners to learn to program, although interface could be a little "harsh".

Lua ​​may be a powerful language but it is simple, and that is why it is generally not the favorite option when developing "serious" programs, for these generally C++, C#, Java, or some well-known engine is used. so don't expect many proffessional developers doing AAA applications in TIC-80, i see is mostly used for experiments, amateur programs or little games.

I am trying to learn to code by programming games using TIC-80, I find it more motivating this way. As I am writing, I realize that the program must perform many repetitive tasks, which I group into functions. Many of these functions can be common between several projects, such as creating objects, drawing sprites or reading inputs, which are grouped into libraries (I even published a "library" cartridge to read the joystick vectorially, see my profile https://tic80.com/dev?id=8725 ).

I think that grouping and reusing code is an important part of development: it keeps ideas organized, saves time, facilitates abstract thinking, and in general makes the whole process easier. But when you can't and you have to keep all the code in a single view, fluid development becomes hard.

That's why I've started several issues asking for code editor in tabs, including multiple files or using external IDEs, or enhance website in browsing projects; all of them are related to the same idea of ​​grouping functions/objects into libraries and reusing them abstractly. I think letting TIC-80 allow development in this way is essential to help beginners in learning (like me) and to experts to develop more agile, please get involved in letting use `require()`, importing code, and/or external ides on standard release.

**Comment by oblerion** [comment]
> please get involved in letting use require(), importing code, and/or external ides on standard release.

I'm an engine developer (C / lua).
And it's complex to use more than one script.

Why is that?
- 1) You have memory space for a single cartridge,
        if you import too many scripts, it's easy to saturate.
- 2) If a player modifies your library.
        You have no control over what he can change in your game.
        Lua can write/read files, which can be dangerous.
- 3) how do external script interact with export game ?

The TIC-80 needs more security and a dynamic memory cartridge for that.

Or you need to make a custom require() to import functions/tables with limited Lua functionality.
And test the space for the cartridge code.

It's huge edit for fantasy computer.

**Comment by oblerion** [comment]
> I wrote it in Spanish (I am Chilean) and translated mostly with Google Translate
it's more powerful than google => https://www.deepl.com

**Comment by atesin** [comment]
> > please get involved in letting use require(), importing code, and/or external ides on standard release.
> 
> I'm an engine developer (C / lua). And it's complex to use more than one script.
> 
> Why is that?
> 
>     * 1. You have memory space for a single cartridge,
>          if you import too many scripts, it's easy to saturate.
> 
>     * 2. If a player modifies your library.
>          You have no control over what he can change in your game.
>          Lua can write/read files, which can be dangerous.
> 
>     * 3. how do external script interact with export game ?
> 
> 
> The TIC-80 needs more security and a dynamic memory cartridge for that.
> 
> Or you need to make a custom require() to import functions/tables with limited Lua functionality. And test the space for the cartridge code.
> 
> It's huge edit for fantasy computer.

i agree with you mostly, but i mean in development stage... in release obviously we have to merge all code inside cart

that is why i also thought in your tic80 package for zerobrane ( https://github.com/oblerion/ZeroBrane-Studio-with-TIC-80 ) in the chance to merge all project lua files before import them in cartridge... something like:

- cat mylib1.lua >> temp.lua
- cat mylib2.lua >> temp.lua
- cat maincode.lua >> temp.lua
- tic80.exe --cmd="load mycart & import temp.lua & save & run"
- rm -f temp.lua

**Comment by technomancy** [comment]
If you want to do game development with multiple modules and external libraries, you're probably better off using [love2d](https://love2d.org) instead of TIC-80 because TIC-80 is explicitly designed for you to do everything from scratch. The constraints are a core part of what makes TIC-80 special and different from alternatives.

**Comment by atesin** [comment]
yes, i want tic80 for learning lua and the basics, then when i reach some level of skillfullness maybe i go for love2d ... but for now i know almost nothing of lua, i feel love2d too big for me, that is what i imagine organize the code on files to ease the developing and learning process
